homeassistant/components/restore_state.py

```diff
@@ -0,0 +1,86 @@
+"""
+Support for restoring entity states on startup.
+
+Component that records all events and state changes. Allows other components
+to query this database.
+
+For more details about this component, please refer to the documentation at
+https://home-assistant.io/components/recorder/
+"""
+import logging
+from datetime import timedelta
+
+import voluptuous as vol
+
+from homeassistant.core import HomeAssistant, split_entity_id
+from homeassistant.const import (
+    EVENT_HOMEASSISTANT_START, ATTR_RESTORED_STATE)
+from homeassistant.components.history import (
+    CONF_DOMAINS, CONF_EXCLUDE, CONF_ENTITIES, CONF_INCLUDE,
+    Filters, get_states, last_recorder_run)
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.typing import ConfigType
+import homeassistant.util.dt as dt_util
+
+DOMAIN = 'restore_state'
+
+DEPENDENCES = ['recorder']
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: cv.FILTER_SCHEMA,
+}, extra=vol.ALLOW_EXTRA)
+
+_LOGGER = logging.getLogger(__name__)
+
+_ALWAYS_OVERWRITE = (
```

Comment:
We should not get `NotImplemetendErrors`. The `added_to_hass` method will be stubbed out in class `Entity`. Each component ABC (ie `Light`) will implement a stub `restore_state` if it decides to implement this logic.

This logic has to live in the components and cannot just pass on the attributes because platforms do not know how their properties are getting represented in the state machine. They only provide the properties and HASS does the rest. We should not unnecessarily couple the two.